aodndata/soop/soop_co2.py

```diff
@@ -0,0 +1,80 @@
+from aodncore.pipeline import HandlerBase, PipelineFile, PipelineFilePublishType, PipelineFileCheckType, FileType
+from aodncore.pipeline.exceptions import InvalidInputFileError
+import soop_co2_nrt_nc_generator
+from soop_co2_classifier import SoopCo2FileClassifier
+import re
+import os
+
+class SoopCo2Handler(HandlerBase):
+    """
+    Handler for data from IMOS SOOP CO2(DM, RT) and Future Reef Map Project.
+    It handles the following file types:
+         * ZIP;
+         * NetCDF;
+         * TXT;
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(SoopCo2Handler, self).__init__(*args, **kwargs)
+        self.allowed_extensions = ['.nc', '.txt', '.zip']
+
+    def preprocess(self):
+        """ Preprocessing for NRT and DM files
+           - NRT: generate a NetCDF files based on input text file.
+             Set the input file publish_type property to 'archive'
+           - DM file collection: update the check_type and publish_type properties for non-NetCDF files.
+             These files are not checked or harvested, but uploaded to S3
+
+        """
+        # Delayed mode file submitted as a zip archive
+        if self.file_extension == '.zip':
+            nc_file = self.file_collection.filter_by_attribute_id('file_type', FileType.NETCDF)
+            if len(nc_file) != 1:
+                raise InvalidInputFileError(
+                    "Expecting one netCDF file in ZIP archive '{zip}'".format(zip=os.path.basename(self.input_file))
+                )
+
+            # first process the NetCDF file to set the destination path for the file collection
+            nc = nc_file[0]
+            set_nc_dest_path = SoopCo2FileClassifier.dest_path(nc)
+            nc.dest_path = os.path.join(set_nc_dest_path, nc.name)
+            nc.publish_type = PipelineFilePublishType.HARVEST_UPLOAD
+            nc.check_type = PipelineFileCheckType.NC_COMPLIANCE_CHECK
+
+            # SOOP-CO2 DM and FRMAP .txt,.pdf or/and .xml files. Set check type to NONEMPTY and publish type to UPLOAD_ONLY
+            non_nc_files = self.file_collection.filter_by_attribute_id('check_type', PipelineFileCheckType.NO_ACTION)
```

Comment:
The "check type" isn't the best way to differentiate files by type. There isn't a specific "filter" method for this exact use case, but you could also get a collection of non-NetCDF files like this:

```python
non_nc_files = PipelineFileCollection(f for f in self.file_collection if f.file_type is not FileType.NETCDF)
```